File: Grind75_1/week3/3Sum.py
"""
https://leetcode.com/problems/3sum/
"""

class Solution:

    def threeSumMap(self, nums: List[int]) -> List[List[int]]:
        
        nums.sort()

        lenOfNums = len(nums)

        valueMap = {}
        
        retList = []

        # fill map with all nums and indexes, if dups the latest index is saved
        for x in range(lenOfNums):
            valueMap[nums[x]] = x
        
        x = 0
        while (x < lenOfNums - 2):
            # lock first value
            # -2 due to nested loop index y = x + 1 and valid 3rd value is > than index y

            if (nums[x] > 0):
                # since sorted asc, there will be no negative numbers after to reduce to sum of 0
                break

            y = x + 1
            while (y < lenOfNums - 1):
                # - 1 due to 3rd valid value is at > index y
                # search for 2nd and 3rd to make sum = 0
                iDiff = (-1) * (nums[x] + nums[y])

                if (iDiff in valueMap and valueMap[iDiff] > y):
                    # if the iDiff exists and it's index is greater than the 2nd index (so that it doesn't use previously evaluated indexes)
                    retList.append([nums[x], nums[y], iDiff])

                # move to the last occurance of it's dup value
                y = valueMap[nums[y]]
                y = y + 1

            # move to the last occurance of it's dup value
            x = valueMap[nums[x]]
            x = x + 1

        return retList


    def threeSum2P(self, nums: List[int]) -> List[List[int]]:

        lenOfNums = len(nums)

        if (lenOfNums < 3):
            return []

        nums.sort()

        if (nums[0] > 0):
            return []

        retList = []

        x = 0
        while (x < lenOfNums - 2):
            if (nums[x] > 0):
                break

            # if the locked value has already been evaluated, move up until next value appears.
            if (x > 0 and nums[x] == nums[x - 1]):
                x = x + 1
                continue

            left = x + 1
            right = lenOfNums - 1
            
            while (left < right):
                iDiff = nums[x] + nums[left] + nums[right]
                #print (str(x) + " : " + str(left) + ", " + str(right))
                if (iDiff > 0):
                    # need a less positive value, move right down
                    right = right - 1
                elif (iDiff < 0):
                    # need more positive value
                    left = left + 1
                else:
                    # 0
                    retList.append([nums[x], nums[left], nums[right]])

                    leftNum = nums[left]
                    rightNum = nums[right]

                    while (left < right and nums[left] == leftNum):
                        # keep moving if dup
                        left = left + 1

                    while (left < right and nums[right] == rightNum):
                        right = right - 1

            x = x + 1

        return retList


    def threeSum(self, nums: List[int]) -> List[List[int]]:
        #return self.threeSumMap(nums)
        return self.threeSum2P(nums)

File: Grind24_Arrays/3-sum.js
// https://leetcode.com/problems/3sum/description/

/**
sort the nums in non-descending order

let i = 0, as the first value of the potential triplet

while (i < nums.length - 2) {
    l = i + 1
    r = nums.length - 1

    while (l < r) 
        sum = nums[i] + nums[l] + nums[r]
        if (sum === 0) {
            res.push([nums[i], nums[l], nums[r]])

            move l forward until no dup of prev
            move r backward until no dup of prev

        else if sum > 0
            need a less negative
            l += 1
        else 
            r -= 1
    
    move i forward until not the same as prev
}

return res

- Time: O(n^2)
- Space: O(n)   n for result
 */

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    if (nums.length < 3) {
        return []
    }

    nums.sort((a, b) => {return a - b})
    const res = new Array()

    let i = 0
    while (i < nums.length - 2) {
        let l = i + 1
        let r = nums.length - 1

        while (l < r) {
            const sum = nums[i] + nums[l] + nums[r]
            if (sum === 0) {
                res.push([nums[i], nums[l], nums[r]])

                let prev = nums[l]
                l += 1
                while (l < r && prev === nums[l]) {
                    l += 1
                }

                prev = nums[r]
                r -= 1
                while (l < r && prev === nums[r]) {
                    r -= 1
                }
            } else if (sum < 0) {
                l += 1
            } else {
                r -= 1
            }
        }
        let prev = nums[i]
        i += 1
        while (i < nums.length - 2 && prev === nums[i]) {
            i += 1
        }
    }

    return res
};